<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Telofinder &mdash; Telofinder 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/plot_directive.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/documentation_options.js?v=8d563738"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            Telofinder
          </a>
              <div class="version">
                Current version: 1.0.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Telofinder</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#positional-arguments">Positional arguments</a></li>
<li><a class="reference internal" href="#optional-arguments">Optional arguments</a></li>
<li><a class="reference internal" href="#help">Help</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-telofinder.telofinder">Reference</a><ul>
<li><a class="reference internal" href="#telofinder.telofinder.base_compos"><code class="docutils literal notranslate"><span class="pre">base_compos()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.classify_telomere"><code class="docutils literal notranslate"><span class="pre">classify_telomere()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.compute_metrics"><code class="docutils literal notranslate"><span class="pre">compute_metrics()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.count_polynuc_occurence"><code class="docutils literal notranslate"><span class="pre">count_polynuc_occurence()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.export_results"><code class="docutils literal notranslate"><span class="pre">export_results()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.get_consecutive_groups"><code class="docutils literal notranslate"><span class="pre">get_consecutive_groups()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.get_entropy"><code class="docutils literal notranslate"><span class="pre">get_entropy()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.get_polynuc"><code class="docutils literal notranslate"><span class="pre">get_polynuc()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.get_strain_name"><code class="docutils literal notranslate"><span class="pre">get_strain_name()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.get_telomere_in_telomeric_reads"><code class="docutils literal notranslate"><span class="pre">get_telomere_in_telomeric_reads()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.output_dir_exists"><code class="docutils literal notranslate"><span class="pre">output_dir_exists()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.parse_arguments"><code class="docutils literal notranslate"><span class="pre">parse_arguments()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.run_on_fasta_dir"><code class="docutils literal notranslate"><span class="pre">run_on_fasta_dir()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.run_on_single_fasta"><code class="docutils literal notranslate"><span class="pre">run_on_single_fasta()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.run_on_single_seq"><code class="docutils literal notranslate"><span class="pre">run_on_single_seq()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.run_telofinder"><code class="docutils literal notranslate"><span class="pre">run_telofinder()</span></code></a></li>
<li><a class="reference internal" href="#telofinder.telofinder.sliding_window"><code class="docutils literal notranslate"><span class="pre">sliding_window()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Telofinder</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Telofinder</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="telofinder">
<h1>Telofinder<a class="headerlink" href="#telofinder" title="Link to this heading">¶</a></h1>
<p>A python package to determine the location and the size of telomeric repeats (both terminal and internal) from genome assemblies. This version of the program only works for yeast genomes (telomere sequence TG<sub>1-3</sub>)</p>
<dl class="simple">
<dt>Telomere detection is based on calculation in a 20 bp sliding window of the following two metrics:</dt><dd><ul class="simple">
<li><p>DNA sequence entropy &lt; entropy_threshold (default=0.8)</p></li>
<li><p>proportion of polynucleotides (default_list = [&quot;CC&quot;, &quot;CA&quot;, &quot;AC&quot;]) &gt;  polynuc_threshold (default=0.8)</p></li>
</ul>
</dd>
</dl>
<img alt="_images/polynuc_entropy_8kb.png" src="_images/polynuc_entropy_8kb.png" />
<img alt="_images/zoom_term.png" src="_images/zoom_term.png" />
<img alt="_images/zoom_int.png" src="_images/zoom_int.png" />
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Link to this heading">¶</a></h1>
<p>You need to have bedtools in your PATH for telofinder to work.</p>
<p>Install (mini)conda:</p>
<blockquote>
<div><p>see <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/">https://docs.conda.io/projects/conda/en/latest/user-guide/install/</a></p>
</div></blockquote>
<p>Configure channels and add dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>conda-forge<span class="w"> </span>bioconda
$<span class="w"> </span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>telofinder<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.10
$<span class="w"> </span>conda<span class="w"> </span>activate<span class="w"> </span>telofinder
$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>bedtools
</pre></div>
</div>
<p>install the telofinder package:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/GillesFischerSorbonne/telofinder.git
$<span class="w"> </span><span class="nb">cd</span><span class="w"> </span>telofinder
$<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>.
</pre></div>
</div>
</section>
<section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Link to this heading">¶</a></h1>
<p>Using the conda environment created during Installation:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>conda<span class="w"> </span>activate<span class="w"> </span>telofinder
$<span class="w"> </span>telofinder<span class="w"> </span><span class="o">[</span>fasta_path<span class="o">]</span>
</pre></div>
</div>
<section id="positional-arguments">
<h2>Positional arguments<a class="headerlink" href="#positional-arguments" title="Link to this heading">¶</a></h2>
<blockquote>
<div><dl class="simple">
<dt>fasta_path</dt><dd><p>path to a single (multi)fasta file or to a directory containing multiple fasta files</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="optional-arguments">
<h2>Optional arguments<a class="headerlink" href="#optional-arguments" title="Link to this heading">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-f</span>, <span class="option">--force</span></kbd></dt>
<dd><p>automatically replace the output 'telofinder_results' directory if present</p>
</dd>
<dt><kbd><span class="option">-e</span>, <span class="option">--entropy_threshold</span></kbd></dt>
<dd><p>entropy threshold for telomere prediction,
default = 0.8</p>
</dd>
<dt><kbd><span class="option">-n</span>, <span class="option">--polynuc_threshold</span></kbd></dt>
<dd><p>poly-nucleotide threshold for telomere prediction,
default = 0.8</p>
</dd>
<dt><kbd><span class="option">-s</span>, <span class="option">--nb_scanned_nt</span></kbd></dt>
<dd><p>total number of nucleotides scanned for telomere detection, starting from each chromosome extremity. If set to -1, the whole chromosome sequences will be scanned, default = 20 000 bp</p>
</dd>
<dt><kbd><span class="option">-t</span>, <span class="option">--threads</span></kbd></dt>
<dd><p>number of threads to use, default = 1</p>
</dd>
<dt><kbd><span class="option">-r</span>, <span class="option">--raw</span></kbd></dt>
<dd><p>outputs the raw dataframe (raw_df.csv) containing the values of all sliding windows</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="help">
<h2>Help<a class="headerlink" href="#help" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>python<span class="w"> </span>telofinder.py<span class="w"> </span>--help
</pre></div>
</div>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">¶</a></h2>
<p>Telofinder outputs a directory called <code class="docutils literal notranslate"><span class="pre">telofinder_results</span></code> including 2 csv and 2 bed files containing the telomere calls and their coordinates, either as raw output or after merging consecutive calls</p>
</section>
</section>
<section id="module-telofinder.telofinder">
<span id="reference"></span><h1>Reference<a class="headerlink" href="#module-telofinder.telofinder" title="Link to this heading">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.base_compos">
<span class="sig-name descname"><span class="pre">base_compos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#base_compos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.base_compos" title="Link to this definition">¶</a></dt>
<dd><p>Counts the number of a given base in a sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> -- fasta sequence</p></li>
<li><p><strong>base</strong> -- base to count in the sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the number of that base in the sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.classify_telomere">
<span class="sig-name descname"><span class="pre">classify_telomere</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interval_chrom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrom_len</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#classify_telomere"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.classify_telomere" title="Link to this definition">¶</a></dt>
<dd><p>From a list of tuples obtained from get_consecutive_groups, identify if
interval corresponds to terminal or interal telomere</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.compute_metrics">
<span class="sig-name descname"><span class="pre">compute_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynucleotide_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['AC',</span> <span class="pre">'CA',</span> <span class="pre">'CC']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#compute_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.compute_metrics" title="Link to this definition">¶</a></dt>
<dd><p>Compute entropy and polynucleotide proportion in the sequence window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> -- sliding window</p></li>
<li><p><strong>polynucleotide_list</strong> -- a list of polynucleotides, default value is [&quot;AC&quot;, &quot;CA&quot;, &quot;CC&quot;]</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a dictionary of entropy and polynucleotide proportion of the sequence window</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.count_polynuc_occurence">
<span class="sig-name descname"><span class="pre">count_polynuc_occurence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sub_window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynucleotide_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#count_polynuc_occurence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.count_polynuc_occurence" title="Link to this definition">¶</a></dt>
<dd><p>Define presence of polynucleotide in the window.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sub_window</strong> -- the sequence of a sub_window</p></li>
<li><p><strong>polynuleotide_list</strong> -- a list of polynucleotides. Note that all polynucleotides must be of the same size</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a boolean for the presence of the sub_window in the polynucleotide list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.export_results">
<span class="sig-name descname"><span class="pre">export_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">telom_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merged_telom_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#export_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.export_results" title="Link to this definition">¶</a></dt>
<dd><p>Produce output table files</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.get_consecutive_groups">
<span class="sig-name descname"><span class="pre">get_consecutive_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_chrom</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#get_consecutive_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.get_consecutive_groups" title="Link to this definition">¶</a></dt>
<dd><p>From the raw dataframe get start and end of each telomere window.
Applied to detect start and end of telomere in nucleotide positions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.get_entropy">
<span class="sig-name descname"><span class="pre">get_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#get_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.get_entropy" title="Link to this definition">¶</a></dt>
<dd><p>Calculate the entropy of the window DNA sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>window</strong> -- sliding window</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>entropy value of the sequence window</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.get_polynuc">
<span class="sig-name descname"><span class="pre">get_polynuc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynucleotide_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#get_polynuc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.get_polynuc" title="Link to this definition">¶</a></dt>
<dd><p>get the propbortion of polynuceotides in the window</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window</strong> -- sliding window</p></li>
<li><p><strong>polynucleotide_list</strong> -- a list of polynucleotides. Note that all polynucleotides must be of the same size</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>total polynucleotide proportion in the sliding window</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.get_strain_name">
<span class="sig-name descname"><span class="pre">get_strain_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#get_strain_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.get_strain_name" title="Link to this definition">¶</a></dt>
<dd><p>Function to get the strain name from the name of the fasta file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> -- path of fasta file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>sequence name</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.get_telomere_in_telomeric_reads">
<span class="sig-name descname"><span class="pre">get_telomere_in_telomeric_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bam_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">telo_df_merged</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynuc_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_scanned_nt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#get_telomere_in_telomeric_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.get_telomere_in_telomeric_reads" title="Link to this definition">¶</a></dt>
<dd><p>Extract telomeric reads from a bam file corresponding to telomere detected
and reported in telo_df_merged</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam_file</strong> -- An indexed bam alignment file.</p></li>
<li><p><strong>telo_df_merged</strong> -- Merged DataFrame with telomeric informations (from one of the run_telofinder functions)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.output_dir_exists">
<span class="sig-name descname"><span class="pre">output_dir_exists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#output_dir_exists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.output_dir_exists" title="Link to this definition">¶</a></dt>
<dd><p>Function to test if the 'telofinder_results' output diretory already exists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>force</strong> -- overwrites the output directory otherwise exits program --force or -f</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>nothing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.parse_arguments">
<span class="sig-name descname"><span class="pre">parse_arguments</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#parse_arguments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.parse_arguments" title="Link to this definition">¶</a></dt>
<dd><p>Function to parse and reuse the arguments of the command line</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta_path</strong> -- path to a single fasta file or to a directory containing multiple fasta files</p></li>
<li><p><strong>force</strong> -- force optional, overwrites the output directory otherwise exits program, optional</p></li>
<li><p><strong>entropy_threshold</strong> -- optional, default = 0.8</p></li>
<li><p><strong>polynuc_threshold</strong> -- optional, default = 0.8</p></li>
<li><p><strong>nb_scanned_nt</strong> -- number of scanned nucleotides at each chromosome end, optional, default = 20 000</p></li>
<li><p><strong>threads</strong> -- Number of threads to use. Multithreaded calculations currently occurs at the level of sequences within a fasta file.&quot;</p></li>
<li><p><strong>raw</strong> -- Outputs raw_df.csv containing the values of all sliding windows</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>parser arguments</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.run_on_fasta_dir">
<span class="sig-name descname"><span class="pre">run_on_fasta_dir</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_dir_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynuc_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_scanned_nt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#run_on_fasta_dir"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.run_on_fasta_dir" title="Link to this definition">¶</a></dt>
<dd><p>Run iteratively the telemore detection algorithm on all fasta files in a directory</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>fasta_dir</strong> -- path to fasta directory</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a tuple of df, telo_df and telo_df_merged</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.run_on_single_fasta">
<span class="sig-name descname"><span class="pre">run_on_single_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynuc_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_scanned_nt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#run_on_single_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.run_on_single_fasta" title="Link to this definition">¶</a></dt>
<dd><p>Run the telomere detection algorithm on a single fasta file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>fasta_path</strong> -- path to fasta file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a tuple of df, telo_df and telo_df_merged</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.run_on_single_seq">
<span class="sig-name descname"><span class="pre">run_on_single_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_record</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynuc_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_scanned_nt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#run_on_single_seq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.run_on_single_seq" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.run_telofinder">
<span class="sig-name descname"><span class="pre">run_telofinder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polynuc_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_thres</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_scanned_nt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threads</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#run_telofinder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.run_telofinder" title="Link to this definition">¶</a></dt>
<dd><p>Run telofinder on a single fasta file or on a fasta directory</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="telofinder.telofinder.sliding_window">
<span class="sig-name descname"><span class="pre">sliding_window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/telofinder/telofinder.html#sliding_window"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#telofinder.telofinder.sliding_window" title="Link to this definition">¶</a></dt>
<dd><p>Apply a sliding window of length = size to a sequence from start to end</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> -- fasta sequence</p></li>
<li><p><strong>start</strong> -- starting coordinate of the sequence</p></li>
<li><p><strong>end</strong> -- ending coordinate of the sequence</p></li>
<li><p><strong>size</strong> -- size of the sliding window</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>the coordinate and the sequence of the window</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Link to this heading">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright XX, 2020.
      <span class="lastupdated">Last updated on Oct 20, 2023.
      </span></p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>