---
title: "**PHENOVAR telomere length**"
author: "Gilles Fischer"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: yes
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  powerpoint_presentation:
    slide_level: 2
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
subtitle: Comparison between Canu 2.0 and Smart_de_novo scaffolded assemblies
font-family: Oswald
transition: linear
editor_options: 
  chunk_output_type: inline
---



```{r settings, include=FALSE, echo=FALSE, eval=TRUE}

options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/scaffolding',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

```



```{r libraries, echo=FALSE, eval=TRUE}
requiredLib <- c("knitr", "viridis", "hrbrthemes", "ggplot2", "stringi", "dplyr", "hexbin", "RColorBrewer", "hrbrthemes")
for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}

```



```{r load data, echo=FALSE, eval=TRUE}

### load python telomere data:
## on phenovar strains
tel_canu2 <- read.table("/Users/gilles/Documents/DUBii/Projet/results/Telomere_length/Refined_final_set/Ragout_trim/canu2_telom_length.csv",header = T, stringsAsFactors = F)

tel_Sdn <- read.table("/Users/gilles/Documents/DUBii/Projet/results/Telomere_length/Refined_final_set/Ragout_trim/sdn_telom_length.csv",header = T, stringsAsFactors = F)

## load phenovar strain information
strains <- read.table(file = "~/Documents/DUBii/Projet/data/phenovar_strains/117_strains_phenovar.csv", header = TRUE, sep = ";", stringsAsFactors = FALSE, fill = TRUE, na.strings=c("","NA"))

```
### 
```{r number of genome}
strain_list <- unique(tel_canu2$Strain)
tot_genome_canu2 <- length(strain_list)
#tot_genome_canu2
#tot_genome_sdn
```

## Data source: PHENOVAR

- number of assembled genomes: `r tot_genome_canu2`
- scaffolded assemblies are on BiG-Daddy 134.157.11.12 in
  - "/home/samuel/1.phenovar_fq/FINAL_PILON_POLISHED/canu2/refined_final_set/ragout_trim/"
  - also available from the symbolic links ln -ls in ~/Documents/Phenovar_genomes:
    - Canu2_genomes
    - Sdn_genoes
- results are in the dataframes: "canu2_telom_length.csv" and "sdn_telom_length.csv"
- Telomere sizes were obtained running the python script '00_analyze_telom_length-v2.py' using the following criteria:
  - cumulative size of sliding windows of 20 bp from each contig end on the 5' to 3' strand:
    - at most 3 G or T in the 20 bp window: #C + #A>= 17 (telomere sequence (C1-3A)) 
    -  and #C >= 8 and #A" >= 3
    - and not 4 "A" in a row ("AAA" is allowed though)
  - an offset of up to 1500 bases of non-telomeric sequence (ie #C + #A < 19/20) is allowed at both 5' end of the contigs before the presence of the telomere sequence itself. The value of the offset is reported in the dataframe  


## 1 - Distribution of telomere length at contig ends:


```{r distribution telomere length}
hist(tel_canu2$Telom_length, breaks = 50, xlab = "Telomere length", ylab = "number of ends", col='blue', main="")
hist(tel_Sdn$Telom_length, breaks = 50, xlab = "Telomere length", ylab = "number of ends", col=rgb(1,0,0,0.7), add=T, main = "")
legend(800, 500, legend=c("Smart de novo", "Canu2"), col=c("red", "blue"), lty=1:1)

print("Statistics for Canu2.0:")
summary(tel_canu2$Telom_length)

print("Statistics for Smart de novo:")
summary(tel_Sdn$Telom_length)

```
### **Conclusions:**

**- The 2 distributions are highly similar**  
**- The Sdn distribution is a little bit more peaked at 300 bp than that of Canu2**


## 2 - Number of contigs with/without telomeres

```{r contigs w/wo telomeres}

par(mfrow=c(1,2))

### Number of contigs with telomere at both ends
##Canu2.0
tel_canu2$Strain_Chr <-
  stri_join(tel_canu2$Strain,tel_canu2$Chromosome,sep="_")
not_2tel_canu2 <- unique(tel_canu2$Strain_Chr[tel_canu2$Telom_length==0])
#length(unique(tel_canu2$Strain_Chr))
double_tel_canu2 = length(unique(tel_canu2$Strain_Chr)) -
   length(not_2tel_canu2)
#double_tel_canu2
##Sdn
tel_Sdn$Strain_Chr <- stri_join(tel_Sdn$Strain,tel_Sdn$Chromosome,sep="_")
not_2tel_Sdn <- unique(tel_Sdn$Strain_Chr[tel_Sdn$Telom_length==0])
#length(unique(tel_Sdn$Strain_Chr))
double_tel_Sdn = length(unique(tel_Sdn$Strain_Chr)) - length(not_2tel_Sdn)
#double_tel_Sdn

### Number of contigs lacking telomere at both ends
##Canu2.0
totTelLength_canu2 <- aggregate(Telom_length~Strain_Chr, data=tel_canu2, FUN=sum)
no_tel_canu2 <- length(totTelLength_canu2$Strain_Chr[totTelLength_canu2$Telom_length==0])
#no_tel_canu2
##Sdn
totTelLength_Sdn <- aggregate(Telom_length~Strain_Chr, data=tel_Sdn, FUN=sum)
no_tel_Sdn <- length(totTelLength_Sdn$Strain_Chr[totTelLength_Sdn$Telom_length==0])
#no_tel_Sdn

## Number of contigs with telomere at only 1 end
##Canu2.0
single_tel_canu2 <- length(totTelLength_canu2$Strain_Chr) - double_tel_canu2 -
  no_tel_canu2
#single_tel_canu2
##Sdn
single_tel_Sdn <- length(totTelLength_Sdn$Strain_Chr) - double_tel_Sdn -
  no_tel_Sdn
#single_tel_Sdn

##barplots
##Canu2.0
barplot(c(double_tel_canu2, single_tel_canu2, no_tel_canu2), names.arg = c("2 telom","1 telom", "0 telom"), main = "Repartition of telomere ends", ylim = c(0, 1600), ylab = "Number of contigs in Canu2", col = "blue")
##Sdn
barplot(c(double_tel_Sdn, single_tel_Sdn, no_tel_Sdn), names.arg = c("2 telom","1 telom", "0 telom"), main = "", ylim = c(0, 1600), ylab = "Number of contigs in Sdn", col=rgb(1,0,0,0.7))

par(mfrow=c(1,1))
```
### **Conclusions:**

**- Same total number of scaffolds in Canu2 and Sdn: 1895 scaffolds (representing 3790 scaffold ends).**  
**- Canu2 79%, 15% and 6% of contigs have 2, 1 and 0 telomeres, respectively.**  
**- Sdn 78%, 19% and 3% of contigs have 2, 1 and 0 telomeres, respectively.**  

## 3 - Presence of offset sequences before telomeres
```{r offsets}

par(mfrow=c(2,2))

### Distribution of the non-telomeric terminal offset sequences and their associated telomere sizes
##Canu2.0
hist(tel_canu2$Offset[tel_canu2$Offset != 0],col='blue',border=F, breaks = 10, xlab = "size (bp)", main = "", xlim = c(0,1500), ylim = c(0,250))
title(main = "Non-telomeric offset sequences", sub = "Canu2")
##Sdn
hist(tel_Sdn$Offset[tel_Sdn$Offset != 0], col=rgb(1,0,0,0.7),border=F, breaks = 10, xlab = "size (bp)", main = "", xlim = c(0,1500), ylim = c(0,250))
title(main = "", sub = "Sdn")

##Canu2.0
hist(tel_canu2$Telom_length[tel_canu2$Offset != 0],col='blue', border=F, breaks = 10, xlab = "size (bp)", main ="", xlim = c(0,800), ylim = c(0,80))
title(main = "Offset-associated telomere sizes", sub = "Canu2")
##Sdn
hist(tel_Sdn$Telom_length[tel_Sdn$Offset != 0], col=rgb(1,0,0,0.7), border=F, breaks = 20, xlab = "size (bp)", main ="", xlim = c(0,800), ylim = c(0,80))
title(main = "", sub = "Sdn")

par(mfrow=c(1,1))
```

### **Conclusions:**

**- Sdn scaffolded assemblies contain many more offset sequences than Canu assemblies**  

## 4 - Number of telomere per strain

```{r telomere per strain}

### filter the contig ends without telomere
##Canu2.0
tel_canu2_filtered <- filter(tel_canu2, Telom_length != 0)
##Sdn
tel_Sdn_filtered <- filter(tel_Sdn, Telom_length != 0)

### count the number of telomere per strain
nb_tel_per_strain_canu2 <- count(tel_canu2_filtered, Strain)
nb_tel_per_strain_Sdn <- count(tel_Sdn_filtered, Strain)

### Hist of the number of telom per strain
hist(nb_tel_per_strain_canu2$n, main = "Number of telomere per strain", xlab="", col = 'blue', breaks = 33, xlim=c(0,35), ylim = c(0,70))
hist(nb_tel_per_strain_Sdn$n, xlab="", breaks = 32, add=T, main = "", col=rgb(1,0,0,0.7))
legend(0, 70, legend=c("Smart de novo", "Canu2"), col=c("red", "blue"), lty=1:1)

## number of strains with 32 telomeres
tel32_canu2 <- filter(nb_tel_per_strain_canu2, nb_tel_per_strain_canu2$n == 32)
tel32_Sdn <- filter(nb_tel_per_strain_Sdn, nb_tel_per_strain_Sdn$n == 32)

## print the name of the strains with more than 32 telomeres
more_32tel_canu2 <- filter(nb_tel_per_strain_canu2, nb_tel_per_strain_canu2$n > 32)
more_32tel_Sdn <- filter(nb_tel_per_strain_Sdn, nb_tel_per_strain_Sdn$n > 32)

## find information on the strain with 33 telomeres
#more_32tel_canu2[order(-more_32tel_canu2$n),]
strain_more32tel <- merge(more_32tel_canu2, strains, by ="Strain")
strain_more32tel <- strain_more32tel[, 1:5]

```
### **Conclusions:**

**- Canu 2.0 assemblies are more complete with `r nrow(tel32_canu2)` strains with 32 telomeres as compared to `r nrow(tel32_Sdn)` strains with Smart de novo**  
**- the strain `r more_32tel_canu2$Strain`, `r strain_more32tel$sequenced_as` `r strain_more32tel$original_Zygosity` has `r more_32tel_canu2$n` telomeres in the Canu2 assembly.** 

## 5 - Comparison Canu2 and Sdn telomere lengths for each chromosme end

```{r comparison canu2 Sdn}
### merge canu2 and Sdn dataset
tel_canu2$Strain_Chr_Side <- paste(tel_canu2$Strain_Chr, "_", tel_canu2$Contig_side, sep = "")
tel_Sdn$Strain_Chr_Side <- paste(tel_Sdn$Strain_Chr, "_", tel_Sdn$Contig_side, sep = "")
canu2_Sdn <- merge(x = tel_canu2, y = tel_Sdn, by.x = "Strain_Chr_Side", by.y = "Strain_Chr_Side")
colnames(canu2_Sdn)[5] <- "canu2_size"
colnames(canu2_Sdn)[11] <- "Sdn_size"
canu2_Sdn$diff_size <- canu2_Sdn$canu2_size - canu2_Sdn$Sdn_size

### add the phenovar dataset origin
colnames(canu2_Sdn)[2] <- "Strain"
canu2_Sdn <- merge(canu2_Sdn, strains[, c("Strain", "sequenced_as", "original_Zygosity")], by.x = "Strain", by.y = "Strain")

### correlation with hexbin plot
# Create data
x <- canu2_Sdn$canu2_size
y <- canu2_Sdn$Sdn_size
## Make the plot
bin<-hexbin(x, y, xbins=45)
my_colors=colorRampPalette(rev(brewer.pal(11,'Spectral')))
plot(bin, main="Comparison telomere length" , colramp=my_colors , legend=F, xlab = "Canu2", ylab = "Sdn")

### colorred correlations between the 2 dataset:
## colored as sequenced
ggplot(canu2_Sdn, aes(x=canu2_size, y=Sdn_size, color=sequenced_as)) +
  geom_point(size = 1) +
  xlab("Mean telomere length Canu2") +
  ylab("Mean telomere length Sdn") +
  ggtitle("Comparison Canu2 vs Sdn") +
  scale_color_hue(l=50, c=115)
## colored as zygosity
ggplot(canu2_Sdn, aes(x=canu2_size, y=Sdn_size, color=original_Zygosity)) +
  geom_point(size = 1) +
  xlab("Mean telomere length Canu2") +
  ylab("Mean telomere length Sdn") +
  ggtitle("Comparison Canu2 vs Sdn") +
  scale_color_hue(l=50, c=115)

### Draw histog of the differences:
hist(x = canu2_Sdn$diff_size, breaks = 80, xlab = "", main = "Difference telomere length (Canu2 - Sdn)", col = "purple")

### boxplot difference telomere length per strain
ggplot(canu2_Sdn[1:948,], aes(x=Strain, y=diff_size, fill=Strain)) +
    geom_boxplot() +
    scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    geom_jitter(color="black", size=0.4, alpha=0.9) + 
    ggtitle("Distribution of telomere length difference per strain") +
    theme(
      legend.position="none",
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(canu2_Sdn[949:1904,], aes(x=Strain, y=diff_size, fill=Strain)) +
    geom_boxplot() +
    scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      legend.position="none",
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(canu2_Sdn[1905:2810,], aes(x=Strain, y=diff_size, fill=Strain)) +
    geom_boxplot() +
    scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      legend.position="none",
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(canu2_Sdn[2811:3650,], aes(x=Strain, y=diff_size, fill=Strain)) +
    geom_boxplot() +
    scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      legend.position="none",
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

## difference between median differences per strain
median_tel_canu2 <- aggregate(tel_canu2$Telom_length, by = list(tel_canu2$Strain), median)
median_tel_Sdn <- aggregate(tel_Sdn$Telom_length, by = list(tel_Sdn$Strain), median)
diff_median <- merge(median_tel_canu2, median_tel_Sdn, by = "Group.1")
colnames(diff_median)[1] <- "Strain"
colnames(diff_median)[2] <- "canu2_median"
colnames(diff_median)[3] <- "Sdn_median"
diff_median$diff <- diff_median$canu2 - diff_median$Sdn_median

ggplot(diff_median, aes(x="117 strains", y=diff)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    xlab("") +
    ylab("Difference between median telomere lengths (canu2 - Sdn)")

```
  
### **Conclusions:**

**- Many discrepencies in telomere length between the 2 assemblies as if telomeres were swapped between chromosomes depending on the assembly tool.**  
**- No clear bias with the ploidy or heterozygosity.**  
**- All strains are affected by the problem but the median of the difference is close to zero in almost all strains.**



### print the 10 smallest and biggest differences:
written in diff_canu2_Sdn.csv
```{r }
## 
sort_diff <- canu2_Sdn[order(canu2_Sdn$diff_size),]
smallest <- head(sort_diff, 10)
biggest <- tail(sort_diff, 10)
extreme <- rbind(smallest, biggest)
diff_canu2_Sdn <- extreme[, c(1, 2, 5, 11, 14)]
diff_canu2_Sdn
write.csv(diff_canu2_Sdn,"~/Documents/DUBii/Projet/results/Telomere_length/Refined_final_set/Ragout_trim/diff_canu2_Sdn", row.names = FALSE)
```

### 6 - Distribution of telomere length per strain in Canu2 assemblies

```{r telomere length per strain, message=FALSE}
### add the phenovar dataset origin to tel_canu2
tel_canu2 <- merge(tel_canu2, strains[, c("Strain", "sequenced_as", "original_Zygosity")], by.x = "Strain", by.y = "Strain")

### plot telomeres per strain
ggplot(tel_canu2[1:966,], aes(x=Strain, y=Telom_length, fill=sequenced_as)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(tel_canu2[967:1910,], aes(x=Strain, y=Telom_length, fill=sequenced_as)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(tel_canu2[1911:2846,], aes(x=Strain, y=Telom_length, fill=sequenced_as)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

ggplot(tel_canu2[2847:3790,], aes(x=Strain, y=Telom_length, fill=sequenced_as)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1))

## dataframes with mean, median and variance of tel_length per strain
moy_tel_canu2 <- aggregate(tel_canu2$Telom_length, by = list(tel_canu2$Strain), mean) 
median_tel_canu2 <- aggregate(tel_canu2$Telom_length, by = list(tel_canu2$Strain), median)
median_tel_canu2 <- merge(median_tel_canu2, strains[, c("Strain", "sequenced_as", "original_Zygosity")], by.x = "Group.1", by.y = "Strain") 
var_tel_canu2 <- aggregate(tel_canu2$Telom_length, by = list(tel_canu2$Strain), var)

```


### **Conclusions:**  
**- Gigascience assemblies lack most telomere sequences**
**- Highly variable telomere length across different strains**
**- very large telomeres in ANM, AVN and BGN  strains (median above 600 bp)**
**- very small telomeres in AIG, BDN and AFI (medians below 200 bp)**
**- the dispersion of telomere lengths varies a lot across genomes**

### 7 - Correlation telomere lenght and variance in Canu2 assemblies

```{r length vs variance, message=FALSE, warning=FALSE}
## filter the contig ends without telomere
tel_filtered <- filter(tel_canu2, Telom_length != 0)
## mean, median and var of tel
moy_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), mean) 
median_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), median)
var_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), var)

## merge the 3 df
moy_med_tel_filtered <- merge(moy_tel_filtered, median_tel_filtered, by.x = "Group.1", by.y = "Group.1")
moy_med_var_tel_filtered <- merge(moy_med_tel_filtered, var_tel_filtered, by.x = "Group.1", by.y = "Group.1")
colnames(moy_med_var_tel_filtered)<- c("Strain","Mean","Median", "Variance")

## plots
ggplot(moy_med_var_tel_filtered, aes(x=Mean, y=Variance)) +
  geom_point()

ggplot(moy_med_var_tel_filtered, aes(x=Median, y=Variance)) +
  geom_point()

```


**Conclusions:**

- The variance of telomere lenght increases with increasing telomere length
  
  
