---
title: "PHENOVAR: Telomere length comparison across dataset"
author: "Gilles Fischer"
date: '`r Sys.Date()`'
output:
  html_document:
    self_contained: yes
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  powerpoint_presentation:
    slide_level: 2
    fig_caption: yes
    fig_height: 5
    fig_width: 7
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
subtitle: 83 homozygous assemblies from Phenovar vs 19 published homozygous assemblies
font-family: Oswald
transition: linear
editor_options: 
  chunk_output_type: inline
---



```{r settings, include=FALSE, echo=FALSE, eval=TRUE}

options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures',
  fig.align = "center", 
  size = "tiny", 
  echo = TRUE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
  comment = "")

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

```



```{r libraries, echo=FALSE, eval=TRUE}
requiredLib <- c("knitr", "viridis", "hrbrthemes", "ggplot2", "stringi", "dplyr", "hexbin", "RColorBrewer", "hrbrthemes", "tidyr")
for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}

```



```{r load data, echo=FALSE, eval=TRUE}

### load python telomere data:
## 83 homozygous phenovar strains
tel_homoz <- read.table("/Users/gilles/Documents/DUBii/Projet/results/Telomere_length/Final_83_homozygous/telom_length_83_homozygous.csv",header = T, stringsAsFactors = F)

## 19 other strains
tel_other <- read.table("/Users/gilles/Documents/DUBii/Projet/results/Telomere_length/Other_19_strains/telom_length_19_others.csv",header = T, stringsAsFactors = F)

## load phenovar strain information
strains <- read.table(file = "/Users/gilles/Documents/DUBii/Projet/data/phenovar_strains/136_strains_phenovar.csv", header = TRUE, sep = ";", stringsAsFactors = FALSE, fill = TRUE, na.strings=c("","NA"))

```

### Data source

- number of assembled genomes: 83 homozygotes in Phenovar and 19 PacBio/Nanopore from 5 other studies (Jenjaroenpun, Yue, Istace, Czaja and Bendixen)
- Genome assemblies are on BiG-Daddy 134.157.11.12 in
  - "/home/samuel/1.phenovar_fq/FINAL_PILON_POLISHED/homozygous_final/"
  - /home/samuel/1.phenovar_fq/other_3rdgen_genomes/all_others/

### Data processing
- Telomere sizes were obtained running the python script '00_analyze_telom_length-v2.py' using the following criteria:
  - cumulative size of sliding windows of 20 bp from each contig end on the 5' to 3' strand:
    - at most 3 G or T in the 20 bp window: #C + #A>= 17 (telomere sequence (C1-3A)) 
    -  and #C >= 8 and #A" >= 3
    - and not 4 "A" in a row ("AAA" is allowed though)
  - an offset of up to 1500 bases of non-telomeric sequence (ie #C + #A < 19/20) is allowed at both 5' end of the contigs before the presence of the telomere sequence itself. The value of the offset is reported in the dataframe  


## 1 - Distribution of telomere length at contig ends:


```{r distribution telomere length}
hist(tel_homoz$Telom_length, breaks = 50, xlab = "Telomere length", ylab = "number of ends", col='blue', main="")
hist(tel_other$Telom_length, breaks = 50, xlab = "Telomere length", ylab = "number of ends", col=rgb(1,0,0,0.7), add=T, main = "")
legend(800, 300, legend=c("19 others", "83 homozygotes"), col=c("red", "blue"), lty=1:1)

print("Statistics for 83 homozygotes:")
summary(tel_homoz$Telom_length)

print("Statistics for 19 others:")
summary(tel_other$Telom_length)

```
### **Conclusions:**

**- Large excess of end devoid of telomere in the set of 19 strains as compared to the Phenovar strains**  



## 2 - Number of contigs with/without telomeres

```{r contigs w/wo telomeres}

par(mfrow=c(1,2))

### Number of contigs with telomere at both ends
##83 homoz
tel_homoz$Strain_Chr <-
  stri_join(tel_homoz$Strain,tel_homoz$Chromosome,sep="_")
not_2tel_homoz <- unique(tel_homoz$Strain_Chr[tel_homoz$Telom_length==0])


#length(unique(tel_homoz$Strain_Chr))
double_tel_homoz = length(unique(tel_homoz$Strain_Chr)) -
   length(not_2tel_homoz)
#double_tel_homoz
##19 others
tel_other$Strain_Chr <- stri_join(tel_other$Strain,tel_other$Chromosome,sep="_")
not_2tel_other <- unique(tel_other$Strain_Chr[tel_other$Telom_length==0])
#length(unique(tel_other$Strain_Chr))
double_tel_other = length(unique(tel_other$Strain_Chr)) - length(not_2tel_other)
#double_tel_other

### Number of contigs lacking telomere at both ends
##83 homoz
totTelLength_homoz <- aggregate(Telom_length~Strain_Chr, data=tel_homoz, FUN=sum)
no_tel_homoz <- length(totTelLength_homoz$Strain_Chr[totTelLength_homoz$Telom_length==0])
#no_tel_homoz
##19 others
totTelLength_other <- aggregate(Telom_length~Strain_Chr, data=tel_other, FUN=sum)
no_tel_other <- length(totTelLength_other$Strain_Chr[totTelLength_other$Telom_length==0])
#no_tel_other

## Number of contigs with telomere at only 1 end
##83 homoz
single_tel_homoz <- length(totTelLength_homoz$Strain_Chr) - double_tel_homoz -
  no_tel_homoz
#single_tel_homoz
##19 others
single_tel_other <- length(totTelLength_other$Strain_Chr) - double_tel_other -
  no_tel_other
#single_tel_other

##barplots
##83 homoz
barplot(c(double_tel_homoz, single_tel_homoz, no_tel_homoz), names.arg = c("2 telom","1 telom", "0 telom"), main = "Repartition of telomere ends", ylim = c(0, 1600), ylab = "Number of contigs in 83 homozygotes", col = "blue")
##19 others
barplot(c(double_tel_other, single_tel_other, no_tel_other), names.arg = c("2 telom","1 telom", "0 telom"), main = "", ylim = c(0, 1600), ylab = "Number of contigs in 19 others", col=rgb(1,0,0,0.7))

par(mfrow=c(1,1))
```
### **Conclusions:**

**- Phenovar assemblies much more complete**  
 

## 3 - Presence of offset sequences before telomeres
```{r offsets}

par(mfrow=c(2,2))

### Distribution of the non-telomeric terminal offset sequences and their associated telomere sizes
##83
hist(tel_homoz$Offset[tel_homoz$Offset != 0],col='blue',border=F, breaks = 10, xlab = "size (bp)", main = "", xlim = c(0,1500), ylim = c(0,250))
title(main = "Non-telomeric offset sequences", sub = "83 Homozygotes")
##19
hist(tel_other$Offset[tel_other$Offset != 0], col=rgb(1,0,0,0.7),border=F, breaks = 10, xlab = "size (bp)", main = "", xlim = c(0,1500), ylim = c(0,250))
title(main = "", sub = "19 others")

##83
hist(tel_homoz$Telom_length[tel_homoz$Offset != 0],col='blue', border=F, breaks = 10, xlab = "size (bp)", main ="", xlim = c(0,800), ylim = c(0,80))
title(main = "Offset-associated telomere sizes", sub = "83 Homozygotes")
##other
hist(tel_other$Telom_length[tel_other$Offset != 0], col=rgb(1,0,0,0.7), border=F, breaks = 20, xlab = "size (bp)", main ="", xlim = c(0,800), ylim = c(0,80))
title(main = "", sub = "19 others")

par(mfrow=c(1,1))
```

### **Conclusions:**

**- other scaffolded assemblies contain many more offset sequences than Canu assemblies**  




### 4 - Distribution of telomere length per strain

```{r telomere length per strain, fig.height = 21, fig.width = 7, message=FALSE}

### clean strain names in the df
##83
tel_homoz <-  separate(tel_homoz, Strain, c("Strain", "suffix"), sep = "_")
tel_homoz = subset(tel_homoz, select = -suffix)
##19
tel_other <-  separate(tel_other, Strain, c("Strain", "suffix"), sep = "_")
tel_other = subset(tel_other, select = -suffix)

## fuse the 2 df
tel_all <- rbind(tel_homoz, tel_other)

### add the dataset origin to tel_all
tel_all <- merge(tel_all, strains[, c("Strain", "sequenced_as")], by.x = "Strain", by.y = "Strain", all = TRUE)
## sort by origin
tel_all_sorted <- tel_all[order(tel_all$sequenced_as),]

### plot telomeres per strain
ggplot(tel_all, aes(x=Strain, y=Telom_length, fill=sequenced_as)) +
    geom_boxplot() +
    geom_jitter(color="black", size=0.4, alpha=0.9) +
    theme(
      plot.title = element_text(size=11)
    ) +
    xlab("") + theme(axis.text.x=element_text(angle=90, hjust=1)) +
    coord_flip()

#ggsave("distrib_tel_length_sorted.png", plot, width = 7, height = 21)

## dataframes with mean, median and variance of tel_length per strain for homoz
moy_tel_homoz <- aggregate(tel_homoz$Telom_length, by = list(tel_homoz$Strain), mean) 
median_tel_homoz <- aggregate(tel_homoz$Telom_length, by = list(tel_homoz$Strain), median)
median_tel_homoz <- merge(median_tel_homoz, strains[, c("Strain", "sequenced_as", "original_Zygosity")], by.x = "Group.1", by.y = "Strain") 
var_tel_homoz <- aggregate(tel_homoz$Telom_length, by = list(tel_homoz$Strain), var)

```


### **Conclusions:**  
**- Gigascience assemblies lack most telomere sequences**  
**- Highly variable telomere length across different strains**  
**- very large telomeres in ANM, AVN and BGN  strains (median above 600 bp)**  
**- very small telomeres in AIG, BDN and AFI (medians below 200 bp)**  
**- the dispersion of telomere lengths varies a lot across genomes**  

### 5 - Correlation telomere lenght and variance in 83 homozygous assemblies

```{r length vs variance, message=FALSE, warning=FALSE}
## filter the contig ends without telomere
tel_filtered <- filter(tel_homoz, Telom_length != 0)
## mean, median and var of tel
moy_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), mean) 
median_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), median)
var_tel_filtered <- aggregate(tel_filtered$Telom_length, by = list(tel_filtered$Strain), var)

## merge the 3 df
moy_med_tel_filtered <- merge(moy_tel_filtered, median_tel_filtered, by.x = "Group.1", by.y = "Group.1")
moy_med_var_tel_filtered <- merge(moy_med_tel_filtered, var_tel_filtered, by.x = "Group.1", by.y = "Group.1")
colnames(moy_med_var_tel_filtered)<- c("Strain","Mean","Median", "Variance")

## plots
ggplot(moy_med_var_tel_filtered, aes(x=Mean, y=Variance)) +
  geom_point()

ggplot(moy_med_var_tel_filtered, aes(x=Median, y=Variance)) +
  geom_point()

```


**Conclusions:**

**- The variance of telomere lenght increases with increasing telomere length**

## 6 - Number of telomere per strain

```{r telomere per strain per dataset}

### filter the contig ends without telomere
tel_all_filtered <- filter(tel_all, Telom_length != 0)

### count the number of telomere per strain
nb_tel_per_strain <- count(tel_all_filtered, Strain)

### Add information on sequencing dataset
nb_tel_per_strain <- merge(nb_tel_per_strain, strains, by.x = "Strain", by.y = "alt_strain", all.y = FALSE)

### Hist of the number of telom per strain per dataset
ggplot(nb_tel_per_strain, aes(x=n, fill=sequenced_as)) +
    geom_histogram( color="#e9ecef", position = 'identity') +
    theme_ipsum() +
    labs(fill="") +
    xlab("Number of telomere per strain") +
    ylab("Number of strains")

## strains with 32 telomeres
tel32 <- filter(nb_tel_per_strain, nb_tel_per_strain$n == 32)
tel32_stat <- merge(tel32, moy_med_var_tel_filtered, by.x = "Strain", by.y = "Strain", all.y = FALSE)
plot(tel32_stat$Mean~tel32_stat$Median)
plot(tel32_stat$Mean~tel32_stat$Variance)
```
### **Conclusions:**

**- again Phenovar assemblies appear the most complete of all** 
  